Modbus Input plugin

garciaolais opened the pull request:
This adds a new input plugin for Modbus TCP/RTU protocol 


### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

danielnelson commented:
@fdamador Can you review?

Does this supersede #4623?

danielnelson commented:
@garciaolais To get the tests running you will need to run `dep ensure` and commit the results.

fdamador commented:
> @fdamador Can you review?
> 
> Does this supersede #4623?

Yes, I've already reviewed his code except for the Test projects.  Yes.  This new commit from @garciaolais  supersedes my #4623 plugin.

garciaolais commented:
> @garciaolais To get the tests running you will need to run `dep ensure` and commit the results.

Done 👍

fdamador commented:
I've ran the modbust_test.go project and it covers about 77.5% of the code.  I'll get some further test code added so that we can get this up to 100%.  Will review over the weekend.

Pradeep-raj544 commented:
@fdamador @garciaolais @ danielnelson 

 i used this link for modbus input plugin
(https://github.com/fdamador/telegraf/tree/0aca266fd588d4c15c36705b7000125bc3d67748/plugins/inputs/modbus)

i am new to telegraf and influxdb
i followed the bellow step 
i kept my modbus.go and modbus_test.go file in
$GOPATH/src/github.com/influxdata/telegraf/plugin/inputs/modbus/
 
registered the plugin at (i just added  "github.com/influxdata/telegraf/plugins/inputs/modbus" in all.go)
$GOPATH/src/github.com/influxdata/telegraf/plugin/inputs/all/all.go

and run make from $GOPATH/src/github.com/influxdata/telegraf.

it created telegraf binary file in the same directory


and executed telegraf --config telegraf.conf --input-filter modbus --test ,in terminal
its giving error ""E! [telegraf] Error running agent: Error: no inputs found, did you provide a valid config file?""

let me knoow if i missed any step
please help me to solve this issue.
just confirm me weather  we can use this for production 
thanks

garciaolais commented:
> let me knoow if i missed any step
> please help me to solve this issue.
> just confirm me weather we can use this for production
> thanks

@Pradeep-raj544 
follow this steps

1.   run `go get github.com/fdamador/telegraf`
2.  `cd go/src/github.com/`
3.   `mv fdamador/ influxdata`
4.   `cd influxdata/telegraf`
5.   `git checkout input/modbus` 
6.  `make`
7. ` ./telegraf --input-filter modbus config > telegraf.conf`
8.  edit telegraf.conf 

```
# Modbus client
[[inputs.modbus]]
 #TCP
 type = "TCP"
 controller="127.0.0.1"
 port = 1502

 #RTU
 #type = "RTU"
 #controller="/dev/ttyUSB0"
 baudRate = 9600
 dataBits = 8
 parity = "N"
 stopBits = 1

 slaveId = 1
 timeout = 1

  [[inputs.modbus.Registers.HoldingRegisters.Tags]]
   name = "Voltage"
   order = "AB"
   datatype = "FLOAT32"
   scale = "/10"
   address = [
    0
   ]

  [[inputs.modbus.Registers.HoldingRegisters.Tags]]
   name = "Current"
   order ="ABCD"
   datatype = "FLOAT32"
   scale = "/1000"
   address = [
    1,
    2
   ]

  [[inputs.modbus.Registers.HoldingRegisters.Tags]]
    name = "Power"
    order = "ABCD"
    datatype = "FLOAT32"
    scale = "/10"
    address = [
     3,
     4
    ]

  [[inputs.modbus.Registers.HoldingRegisters.Tags]]
    name = "Energy"
    order = "CDAB"
    datatype = "FLOAT32"	
    scale = "/1000"
    address = [
     5,
     6
    ]

  [[inputs.modbus.Registers.HoldingRegisters.Tags]]
    name = "Frequency"
	order = "AB"	
	datatype = "FLOAT32"	
    scale = "/10"
    address = [
     7
    ]

  [[inputs.modbus.Registers.HoldingRegisters.Tags]]
    name = "PowerFactor"
    order = "AB"
    datatype = "FLOAT32"
    scale = "/100"
    address = [
     8
    ]

```
9. if you don't have a controller run this go example

```
package main

import (
	"log"
	"time"
	"fmt"

	"github.com/tbrandon/mbserver"
	"github.com/goburrow/modbus"
)

func main() {
	serv := mbserver.NewServer()
	err := serv.ListenTCP("127.0.0.1:1502")
	if err != nil {
		log.Printf("%v\n", err)
	}
	defer serv.Close()

	handler := modbus.NewTCPClientHandler("localhost:1502")
	// Connect manually so that multiple requests are handled in one session
	err = handler.Connect()
	defer handler.Close()
	client := modbus.NewClient(handler)

	var len uint16 = 10
	_, err = client.WriteMultipleRegisters(0, len, []byte{
		0x08, 0x98, 
		0x00, 0x00, 0x03, 0xE8,
		0x00, 0x00, 0x08, 0x98,	
		0x00, 0x00, 0x00, 0x00,	
		0x01, 0xF4,	
		0x00, 0x64,	
		0x00, 0x00,	
		})
	if err != nil {
		fmt.Printf("%v\n", err)
	}

	results, err := client.ReadHoldingRegisters(0, len)
	if err != nil {
		fmt.Printf("%v\n", err)
	}
	fmt.Printf("Voltage     0x%x\n", results[0:2])
	fmt.Printf("Current     0x%x\n", results[2:6])
	fmt.Printf("Power       0x%x\n", results[6:10])
	fmt.Printf("Energy      0x%x\n", results[10:14])
	fmt.Printf("Frecuency   0x%x\n", results[14:16])
	fmt.Printf("PowerFactor 0x%x\n", results[16:18])
	fmt.Printf("Alarm       0x%x\n", results[18:20])

	// Wait forever
	for {
		time.Sleep(1 * time.Second)
	}
}
```
10. run `./telegraf -config telegraf.conf -input-filter modbus -test`

you will see something like this

```
2019-08-02T13:31:45Z I! Starting Telegraf
> modbus.HoldingRegister,host=DESKTOP-5000 Current=1,Energy=0,Frequency=50,Power=220,PowerFactor=1,Voltage=220
```

Pradeep-raj544 commented:
@garciaolais 

you are the Savior , it worked like charm.
if i found any issues i will get back to you.
Thank you so much for your  valuable time and help @garciaolais

danielnelson commented on plugins/inputs/modbus/modbus_test.go:
```diff
@@ -0,0 +1,261 @@
+		},
+	}
+
+	regTests := []mbTest{}
```
Let's see if we can set this up as a [table test](https://github.com/golang/go/wiki/TableDrivenTests).  Each test or subtest should perform one logical check.

Here is an example from within Telegraf:https://github.com/influxdata/telegraf/blob/master/plugins/outputs/influxdb/http_test.go#L61

danielnelson commented on plugins/inputs/modbus/modbus_test.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+func TestCoils(t *testing.T) {
+	serv := mbserver.NewServer()
```
It could be nice to drop the dependency, and creating a container running mbserver would be helpful for that.  If we go this route though then we should mock the client for the bulk of the tests and covering most code, and only use the container for broad integration tests.

Pradeep-raj544 commented:
@garciaolais @danielnelson
i got plugin only for tcp and rtu , 
any one developed modbus input plugin for RTU over TCP if any please could you share with me.
thank you

danielnelson commented:
I believe we would need to implement it in goburrow/modbus first, here is an open issue I found: https://github.com/goburrow/modbus/issues/37

@Pradeep-raj544 Even though this PR isn't merged yet, can you go ahead and create a new feature request issue here on Telegraf to track RTU over TCP?

garciaolais commented:
@danielnelson , @glinton  

Sorrry for the delay, I will check 👀  all the reviews and I will working 🔧  as soon as possible

Pradeep-raj544 commented:
could you please implement retries in this plugin, i have tested this plugin for 4(1, 2 ,3, 4) devices simultaneously (multiple) ,the plugin is not polling devices properly , it is polling randomly ,it polls only 2 devices for five seconds and after five sec 3 devices the after five sec 4 devices i.e 4th device querying and the response not coming (here its not waiting for response only) , polling continuously without waiting..
and one more issue RTU  not working.. properly
when it is RTU ,its querying but response not showing ...only it querying but not returning any value

ToM-MaS commented:
@garciaolais 
Any Timeline for the finalisation? Just kindly asking. :)
I'd like to know if I can wait until the final verion is out for my energy meter project.

Another question:
The Energy meters made by www.janitza.de update their internal modbus registers every 0.2 seconds. 
When I poll these values each 0.2 seconds over TCP, I want to:
- not loose an updated value because in my setup each value is important
- not read a value twice (or discard a duplicated value)
But because afaik modbus does not provide any synchronisation method. I'm not sure how and where to solve this. Does anyone have an idea?

danielnelson commented:
> not loose an updated value because in my setup each value is important
> not read a value twice (or discard a duplicated value)

As this plugin is polling based solution, I don't it will be able to fulfil those requirements.  To not miss values you would need something that can receive events, and AFAIK Modbus in general does not have a mechanism for this.

You might be able to get close though, if you dial down the `interval` and if we also created a processor to throw away duplicate data, but it wouldn't be guaranteed.

garciaolais commented:
> @garciaolais
> Any Timeline for the finalisation? Just kindly asking. :)

@ToM-MaS  , Today 📆  I started working  🛠  in the reviews ,  sorry for the delay ⏲ 😞 

https://github.com/garciaolais/telegraf/commits/input/modbus_reviews/plugins/inputs/modbus

ToM-MaS commented:
> You might be able to get close though, if you dial down the `interval` and if we also created a processor to throw away duplicate data, but it wouldn't be guaranteed.

I'd like to do it this way. Now the question is where to place this processor? 

- in this modbus input plugin?
- in the used modbus  connector: https://github.com/goburrow/modbus/
- in Influx Capacitor?
- or somewhere else?

glinton commented:
>Now the question is where to place this processor?

ToM-MaS, I assume he was referring to a standalone [processor plugin](https://github.com/influxdata/telegraf#processor-plugins)

glinton commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,595 @@
+	}
+
+	if !m.isInitialized {
+		err := initialization(m)
```
This can be part of the `Init() error` function introduced in #5899. no change required

garciaolais replied:
✔️ done

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,595 @@
+
+	switch u.Scheme {
+	case "tcp":
+		host, port, _err := net.SplitHostPort(u.Host)
```
✔️ done

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,595 @@
+			m.isConnected = true
+			return nil
+		} else {
+			return fmt.Errorf("Not valid protcol [%s] - [%s] ", u.Scheme, m.TransmissionMode)
```
✔️ done

danielnelson commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+	for _, reg := range m.registers {
+		fields = addFields(reg.Fields)
+		// telegraf.Accumulator.AddFields(Name, Fields, Tags, Time)
+		acc.AddFields("modbus."+reg.Type, fields, tags)
```
Add the register type as a tag using snake case: `input_register`.  Have the measurement be just `modbus`.

garciaolais replied:
done ✔️

danielnelson commented on plugins/inputs/modbus/README.md:
```diff
@@ -0,0 +1,82 @@
+   { name = "Start",          address = [0]},   
+   { name = "Stop",           address = [1]},
+   { name = "Reset",          address = [2]},
+   { name = "EmergencyStop",  address = [3]},
```
Can you update the sample configuration to use snake_case for the field/tag names?  This will better match other Telegraf plugins.

garciaolais replied:
done ✔️

fdamador commented:
Wow, Looks like this code finally made it.  Good work everyone.

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+	for i := 0; i < r.NumField(); i++ {
+		f := r.Field(i)
+
+		if f.Type().String() == "[]modbus.fieldContainer" {
```
I have removed the reflect code.

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+	}
+}
+
+func convertEndianness16(o string, b []byte) interface{} {
```
Done ✔️

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+	dataType := []string{"UINT16", "INT16", "UINT32", "INT32", "FLOAT32-IEEE", "FLOAT32"}
+
+	nameEncountered := map[string]bool{}
+	for i := range t {
```
I have implemented this form  `for _, item := range t ` in some cases

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+		if n == cInputRegisters || n == cHoldingRegisters {
+			// search byte order
+			byteOrderEncountered := false
+			for j := range byteOrder {
```
done ✔️

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+			}
+
+			// check scale
+			_, err := strconv.ParseFloat(t[i].Scale, 32)
```
I have removed the string parse

garciaolais commented on plugins/inputs/modbus/modbus.go:
```diff
@@ -0,0 +1,589 @@
+
+			if r.Type == cInputRegisters || r.Type == cHoldingRegisters {
+				for i := 0; i < len(res); i += 2 {
+					rawValues[rr.address+uint16(i)/2] = uint16(res[i])<<8 | uint16(res[i+1])
```
done ✔️

dontsovcmc commented:
@garciaolais Thanks a lot for you work! Perfect! "Two worlds" joined together. 
Is it possible to read Modbus RTU through TCP port with this plugin? It is needed with 485-Ethernet gateways.

mattbakerau commented:
I would like to use this plugin for a project but I am getting the following error when I attempt to use:

2019-12-26T09:46:31Z E! [telegraf] Error running agent: Error parsing /etc/telegraf/telegraf.d/modbus.conf, line 6: (modbus.fieldContainer.Scale) cannot unmarshal TOML string into float32

Using the following simple configuration:

[[inputs.modbus]]
	slave_id = 1
	controller = "tcp://10.1.1.4:502"

	input_registers = [
		{ name = "OAT", byte_order = "AB", data_type = "INT16", scale="1", address = [10]}
	]


We use modbus a lot and I'm happy to test plugin, but cant seem to get past this first hurdle.

garciaolais commented:
> I would like to use this plugin for a project but I am getting the following error when I attempt to use:
> 
> 2019-12-26T09:46:31Z E! [telegraf] Error running agent: Error parsing /etc/telegraf/telegraf.d/modbus.conf, line 6: (modbus.fieldContainer.Scale) cannot unmarshal TOML string into float32
> 
> Using the following simple configuration:
> 
> [[inputs.modbus]]
> slave_id = 1
> controller = "tcp://10.1.1.4:502"
> 
> ```
> input_registers = [
> 	{ name = "OAT", byte_order = "AB", data_type = "INT16", scale="1", address = [10]}
> ]
> ```
> 
> We use modbus a lot and I'm happy to test plugin, but cant seem to get past this first hurdle.

Hi @mattbakerau 

We changed the data type in the last commit [cc4296e](https://github.com/influxdata/telegraf/pull/6154/commits/cc4296eb5040d0dda067090fddb37be261634dd5) . 

You can use a config like that ->

```
input_registers = [
 	{ name = "OAT", byte_order = "AB", data_type = "INT16", scale=1.0, address = [10]}
]
```

PD: Sorry I forget to update the README and the config file

garciaolais commented:
> @garciaolais Thanks a lot for you work! Perfect! "Two worlds" joined together.
@dontsovcmc thanks!

> Is it possible to read Modbus RTU through TCP port with this plugin? It is needed with 485-Ethernet gateways.
Sorry but at this moment support Modbus RTU through TCP it's out of the scope 😞

mattbakerau commented:
| We changed the data type in the last commit cc4296e .

Great, that worked well.   I now have a bunch of data in influxdb.
Browsing this data I have noticed the name from 'controller' is not added in as a tag.  It would be useful to have either this added, or a 'device name' option.

In my case I have 10x devices giving similar data, so the only way to separate the data recorded would be to give each data point a unique name instead of recording the hostname they come from.

For example I have setup:

# device1
[[inputs.modbus]]
	timeout = "5s"
	slave_id = 1
	controller = "tcp://10.1.1.4:502"

        input_registers = [
              { name = "OAT1", byte_order = "AB", data_type = "INT16", scale=0.1, address = [11]}
       ]

# device2
[[inputs.modbus]]
	timeout = "5s"
	slave_id = 1
	controller = "tcp://10.1.1.5:502"

        input_registers = [
                 { name = "OAT2", byte_order = "AB", data_type = "INT16", scale=0.1, address = [11]}
        ]

garciaolais commented:
> In my case I have 10x devices giving similar data, so the only way to separate the data recorded would be to give each data point a unique name instead of recording the hostname they come from.

@mattbakerau I've updated ⬆️  the code and the config file 📁 , I've added a new field **name**

```
 name = "Device"
 
 ## Slave ID - addresses a MODBUS device on the bus
 ## Range: 0 - 255 [0 = broadcast; 248 - 255 = reserved]
 slave_id = 1
```

If you have a chance, please update ⬇️  your repo and test the new feature.

mattbakerau commented:
Thanks for adding that feature.  I have updated my code and it works well.

One thing I have just noticed is I am not sure the scale option is working correctly.  My device reports temperatures with a x10 multiplier.  To get the correct value I used scale = '0.1' and this divided the value, but seems to have lost anything after the decimal point. To confirm I changed to scale = '1.0'.  

Here are a the values being recorded into influxdb (before and after the change to the scale)

> select OAT from modbus
name: modbus
time                OAT
----                ---
1577413720000000000 36
1577413730000000000 36
1577413730000000000 36
1577413730000000000 41
1577413740000000000 36
1577413740000000000 41
1577413740000000000 36
1577413750000000000 41
1577413750000000000 36
1577413750000000000 34
1577413760000000000 364
1577413760000000000 332
1577413760000000000 421
1577413770000000000 332
1577413770000000000 421
1577413770000000000 364
1577413780000000000 421
1577413780000000000 363
1577413780000000000 333
1577413790000000000 364
1577413790000000000 334

garciaolais commented:
> One thing I have just noticed is I am not sure the scale option is working correctly. My device reports temperatures with a x10 multiplier. To get the correct value I used scale = '0.1' and this divided the value, but seems to have lost anything after the decimal point. To confirm I changed to scale = '1.0'.

I think this happens because in your configuration you have a datatype INT16 , you lose precision when you perform the conversion FLOAT to INT.

input_registers = [
             { name = "OAT2", byte_order = "AB", data_type = **"INT16"**, scale=0.1, address = [11]}
    ]

Please check the next example.

Device values

```
address[0] = 1234
address[1] = 1234
address[2] = 1234
address[3] = 1234
```

Config file
```
 input_registers = [   
   { name = "Value-IN16-x10",       byte_order = "AB", data_type = "INT16",   scale=0.1 ,   address = [0]},      
   { name = "Value-INT16-Raw",      byte_order = "AB", data_type = "INT16",   scale=1.0 ,   address = [1]},       
   { name = "Value-FLOAT32-x10",    byte_order = "AB", data_type = "FLOAT32", scale=0.1 ,   address = [2]},      
   { name = "Value-FLOAT32-Raw",    byte_order = "AB", data_type = "FLOAT32", scale=1.0 ,   address = [3]},     
 ]
```

Results

```
Value-IN16-x10    =  123
Value-INT16-Raw   =  1234
Value-FLOAT32-x10 =  123.4
Value-FLOAT32-Raw =  1234.0
```

mattbakerau commented:
That makes sense.  I had falsely assumed the data_type was referring to the modbus data type and not the storage data type.

mattbakerau commented:
After great initial success I decided to add another type of device.  This one works, but after a little while I start getting the following errors logged:

2019-12-27T12:33:30Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '62' does not match request '63'
2019-12-27T12:33:40Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '63' does not match request '64'
2019-12-27T12:33:50Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '64' does not match request '65'
2019-12-27T12:34:00Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '65' does not match request '66'
2019-12-27T12:34:10Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '66' does not match request '67'
2019-12-27T12:34:20Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '67' does not match request '68'
2019-12-27T12:34:30Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '68' does not match request '69'
2019-12-27T12:34:40Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '69' does not match request '70'
2019-12-27T12:34:50Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '70' does not match request '71'
2019-12-27T12:35:00Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '71' does not match request '72'
2019-12-27T12:35:10Z E! [inputs.modbus]: Error in plugin: modbus: response transaction id '72' does not match request '73'

Data is stil be logged though.  I am thinking it might be some sort of timeout, but wondering if there is a good way to find out what the errors actually are referring to?

garciaolais commented:
> Data is stil be logged though. I am thinking it might be some sort of timeout, but wondering if there is a good way to find out what the errors actually are referring to?

@mattbakerau Could you share more information about your settings?

- number of devices
- protocol , TCP or SERIAL
- number of registers
- data type of registers

I'm trying to replicate the error but I can't get it.

mattbakerau commented:
It looks like I have a device that isn't behaving correctly.   Increasing the timeout seems to have helped reduce the number of errors.
Sorry for the noise, I don't think it relates to the plugin.

garciaolais commented:
> We recently switched to using Go modules instead of dep, you can upgrade by removing Gopkg.lock and Gopkg.toml and then running `make && make check`, this should modify go.mod and go.sum and you should be good.

Hi 

Do I need additional steps? 

Maybe update my `Makefile` or something else. 

Because if I remove the Gopkg.* files I get this error...

`could not find project Gopkg.toml, use dep init to initiate a manifest`

fdamador commented:
Perhaps this will help

https://itiskj.hatenablog.com/entry/2018/08/30/101017

fdamador
Sent from XFINITY Connect Mobile App

------ Original Message ------

From: Antonio Garcia
To: influxdata/telegraf
Cc: fdamador, Mention
Sent: January 25, 2020 at 6:21 PM
Subject: Re: [influxdata/telegraf] Modbus Input plugin (#6154)

>
> We recently switched to using Go modules instead of dep, you can upgrade by removing Gopkg.lock and Gopkg.toml and then running make && make check, this should modify go.mod and go.sum and you should be good.
>

Hi

Do I need additional steps?

Maybe update my Makefile or something else.

Because if I remove the Gopkg.* files I get this error...

could not find project Gopkg.toml, use dep init to initiate a manifest

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub (https://github.com/influxdata/telegraf/pull/6154?email_source=notifications&email_token=AAMWYEP2PMQFVQ76TZ6LNDTQ7TXRZA5CNFSM4IF7NJAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ5KFJI#issuecomment-578462373), or unsubscribe (https://github.com/notifications/unsubscribe-auth/AAMWYEJG2EKU67XRQPHXJ33Q7TXRZANCNFSM4IF7NJAA).

danielnelson commented:
@garciaolais Thanks for bearing with me during this transition, perhaps it will help if you merge the master branch?  If that's not all there is to it, do you mind if I push to your branch and fix it?

garciaolais commented:
> @garciaolais  If that's not all there is to it, do you mind if I push to your branch and fix it?

@danielnelson  any help is highly appreciated 👍

serrj-sv commented:
Is it possible to make a conversion to FLOAT32 more neat?
Using latest nightly, with following config
 `  holding_registers = [
    { name = "active_power",   byte_order = "ABCD", data_type = "FLOAT32", scale=0.01, address = [50536,50537]},
  ]
`
I  get following result:
`> modbus,host=influx,name=mon1.pdr1,type=holding_register active_power=99.72999572753906 1581094677000000000`
Note amount of digits after point

serrj-sv commented:
Found one more issue: if host associated with input plugin is down it prevents telegraf service from starting. Service goes into loop with following message:
`E! [telegraf] Error running agent: could not initialize input inputs.modbus: dial tcp 10.0.4.70:502: i/o timeout`

danielnelson commented:
On the float noise, we probably can't improve upon it.  Could be interesting to see what value you get if you set the data type to UINT32 with no scale (just to look at the raw value).  One thing that could possilby be done is a manual setting for precision and we could round the value.  In most cases I would recommend handling this in the tool you are using to display the value.

@garciaolais For the initialization issue, would we be able to remove the `connect` function out of the Init function?

serrj-sv commented:
> Could be interesting to see what value you get if you set the data type to UINT32 with no scale (just to look at the raw value).

`FLOAT32 (0.01) ==> UINT32 (1.0)
99.27999877929688 => 9928
99.79000091552734 => 9979
99.29999542236328 => 9930
99.22999572753906 => 9923
99.16999816894531 => 9917
98.91999816894531 => 9892
`

danielnelson commented:
Oh, yes we can fix this, we just need to convert to float64 before we apply the scale.

garciaolais commented:
Hi

@danielnelson  @serrj-sv  

Today I worked in these two issues [PR-7002](https://github.com/influxdata/telegraf/pull/7002) 

@serrj-sv If you have a chance please test it 👍

serrj-sv commented:
@garciaolais 
Just tested:
- connection issue seems fixed. I'll monitor how it works
- for float conversion I get mixed result: most values are fine now but one in a while I get following:
1581256130000000000 95.33
1581256140000000000 95.55
1581256150000000000 95.97
1581256160000000000 96.15
1581256170000000000 96.75
1581256180000000000 96.82000000000001
1581256190000000000 97.18
1581256200000000000 96.94
1581256210000000000 96.24000000000001
1581256220000000000 96.32000000000001
1581256230000000000 95.76
1581256240000000000 95.48
1581256250000000000 95.63
1581256260000000000 95.83
1581256270000000000 96
1581256280000000000 94.79
1581256290000000000 94.63
- Not an issue but rather an suggestion: possibility to specify multiple hosts per "input" config section (for cases when you collect data from multiple SIMILAR devices). The way it works with [inouts.snmp](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/snmp), for example

garciaolais commented:
> * for float conversion I get mixed result:
>   1581256130000000000 95.33
>   1581256140000000000 95.55
>   1581256150000000000 95.97
>   1581256160000000000 96.15
>   1581256170000000000 96.75
>   1581256180000000000 96.82000000000001
>   1581256190000000000 97.18
>   1581256200000000000 96.94
>   1581256210000000000 96.24000000000001
>   1581256220000000000 96.32000000000001
>   1581256230000000000 95.76
>   1581256240000000000 95.48
>   1581256250000000000 95.63
>   1581256260000000000 95.83
>   1581256270000000000 96
>   1581256280000000000 94.79
>   1581256290000000000 94.63

@serrj-sv  Thanks for your feedback. I think the conversion its the normal result, please check the next example -> [float64-golang](https://play.golang.org/p/byKzJaESIx-)

and this post -> [why-am-i-losing-precision-while-converting-float32-to-float64](https://stackoverflow.com/questions/39642810/why-am-i-losing-precision-while-converting-float32-to-float64)

fbergroth commented:
Hi, maybe I'm missing something, but is `scale` enough for input registers?

I have signals where the values needs to be remapped to a certain range, say [-50, 150], so I would need a way to specify an offset as well (or ideally just (min, max)).`

danielnelson commented:
@fbergroth Can you open a new feature request?

fbergroth commented:
@danielnelson Sure:  #7009

ToM-MaS commented:
Hello everyone,
for testing I use a public demo device of a manufacturer.

The manufacturers manual shows this for the Modbus adress list:
```
Address	Format 	RD/WR 	Unit	Comment
19000 	FLOAT 	RD 	V 	Voltage L1-N
19000 	FLOAT 	RD 	V 	Voltage L1-N
19002 	FLOAT 	RD 	V 	Voltage L2-N
19004 	FLOAT 	RD 	V 	Voltage L3-N
19006 	FLOAT 	RD 	V 	Voltage L1-L2
19008 	FLOAT 	RD 	V 	Voltage L2-L3
19010 	FLOAT 	RD 	V 	Voltage L1-L3
```

I have configured in telegraf.conf:

```
[[inputs.modbus]]
  name = "Device_UMG604"
  slave_id = 1

  ## Timeout for each request
  timeout = "1s"

  # TCP - connect via Modbus/TCP
  controller = "t<delete_this>cp:/<delete_this>/umg<delete_this>604.jani<delete_this>tza.de:5<delete_this>02"
 
 input_registers = [
    { name = "Voltage L1-N (V)",   byte_order = "ABCD",   data_type = "FLOAT32",   scale=1.0,     address = [19000,19001]},
  ]
```

When I click together a query in Chronograph (and modify the query as shown below), the left scale of the graph shows as values: 1.13e+9 Volt or 1.13B Volt. Expected would be something around 230.0 Volt  (float value).

`SELECT "Voltage L1-N (V)" AS "Voltage L1-N (V)" FROM "telegraf"."autogen"."modbus" WHERE time > :dashboardTime: AND time < :upperDashboardTime: AND "name"='Device_UMG604'`

So I had it exported as CSV. It looks like this:

```
"time","modbus.Voltage L1-N (V)"
"2020-04-26T21:38:50.000Z","1130678532"
"2020-04-26T21:38:55.000Z","1130684293"
"2020-04-26T21:39:00.000Z","1130679969"
"2020-04-26T21:39:05.000Z","1130671574"
"2020-04-26T21:39:10.000Z","1130679470"
"2020-04-26T21:39:15.000Z","1130677593"
"2020-04-26T21:39:20.000Z","1130679974"
"2020-04-26T21:39:25.000Z","1130672015"
"2020-04-26T21:39:30.000Z","1130675830"
"2020-04-26T21:39:35.000Z","1130675929"
"2020-04-26T21:39:40.000Z","1130675135"
"2020-04-26T21:39:45.000Z","1130679073"
```


1st question:
It seems the values in InfuxDB are integer instead of float32. Is this expected?
Where can or should I change something to get float values in Chronograph? (e.g. the value for the upper example should be around e.g. 230.0 Volt)

2nd question:
How do I set that my modbus device is queried every 100ms? From the CSV values above it seems that there is a new row each 5 seconds. Not sure where this rythm comes from, does anyone know?!

garciaolais commented:
Hello @ToM-MaS 

I can help you with your first question

> 1st question:
> It seems the values in InfuxDB are integer instead of float32. Is this expected?
> Where can or should I change something to get float values in Chronograph? (e.g. the value for the upper example should be around e.g. 230.0 Volt)

I read the documentation for your device   [UMG-604](https://www.janitza.com/manuals.html?file=files/download/manuals/current/UMG604-PRO/janitza-mal-umg604pro-en.pdf)
![image](https://user-images.githubusercontent.com/14133711/81483373-5b261100-9203-11ea-9773-d26e78a7dd34.png)

Please note two things

1. The byte order in your device
2. Float values are in IEEE-754

So, I made a little test with my device

My Modbus Device
![image](https://user-images.githubusercontent.com/14133711/81483514-7e9d8b80-9204-11ea-85ea-fbe1623a094c.png)

telegraf.conf

{ name = "VL1-N",byte_order = "ABCD",data_type = "FLOAT32-IEEE", scale=1.0, address = [2,1]}

modbus,host=9e88b516fd94,name=Device,type=holding_register 
### VL1-N= 228.78521728515625 ✨ 

PD. You can get the same value with different configuration for example

{ name = "VL1-N-ABCD",
  **byte_order = "ABCD",** 
  data_type = "FLOAT32-IEEE", 
   scale=1.0, 
   **address = [2,1]}**

{  name = "VL1-N-CDAB",
    **byte_order = "CDAB"**,
    data_type = "FLOAT32-IEEE", 
    scale=1.0, 
    **address = [1,2]**
}

ToM-MaS commented:
Hello @garciaolais
the byte order was OK but the "data_type = "FLOAT32-IEEE"" instead of "FLOAT32" gave me sane values.
 So I'm wondering where the data type FLOAT32-IEEE is documented? Are there more data types if this kind?

BTW, I solved my question about the query frequency by adding the line `interval = "100ms"` to the inputs.modbus section:
```
[[inputs.modbus]]
  ## Query interval for each request
  interval = "100ms"
```

garciaolais commented:
> So I'm wondering where the data type FLOAT32-IEEE is documented? Are there more data types of this kind?

Hello @ToM-MaS 

Well, The device manufacturer define the datatype 

You can find a lot of information about it in this link
https://en.wikipedia.org/wiki/Single-precision_floating-point_format

ToM-MaS commented:
Hello @garciaolais,
I think I understand now what you mean. I think I understand what FLOAT32-IEEE means.

The template configuration of the modbus plugin says about supported formats:
`## data_type  - INT16, UINT16, INT32, UINT32, INT64, UINT64, FLOAT32, FLOAT32-IEEE (the IEEE 754 binary representation)`

So "DOUBLE64" and "DOUBLE64-IEEE" or something similar is not supported?

It would be intersting to know what datatypes are accepted in telegraf.conf besides the ones listed above.
Are the accepted datatype formats defined by the language specification of GO or by the source code of Telegraf / InfluxDB or by something else?
